src/Processors/Formats/Impl/IonRowOutputFormat.cpp

```diff
@@ -0,0 +1,323 @@
+        }
+        case TypeIndex::Float64: {
+            writer->writeDouble(assert_cast<const ColumnFloat64 &>(column).getElement(row_num));
+            return;
+        }
+        case TypeIndex::DateTime64: {
+            writer->writeBigUInt(assert_cast<const DataTypeDateTime64::ColumnType &>(column).getDataAt(row_num).data, sizeof(DateTime64));
+            return;
+        }
+        case TypeIndex::Decimal32: {
+            writer->writeDecimal32(assert_cast<const ColumnDecimal<Decimal32> &>(column).getElement(row_num).value);
+            return;
+        }
+        case TypeIndex::Decimal64: {
+            writer->writeDecimal64(assert_cast<const ColumnDecimal<Decimal64> &>(column).getElement(row_num).value);
+            return;
+        }
+        case TypeIndex::Decimal128: {
+            writer->writeDecimal128(assert_cast<const ColumnDecimal<Decimal128> &>(column).getDataAt(row_num).data, sizeof(Int128));
+            return;
+        }
+        case TypeIndex::Decimal256: {
+            writer->writeDecimal256(assert_cast<const ColumnDecimal<Decimal256> &>(column).getDataAt(row_num).data, sizeof(Int256));
+            return;
+        }
+        case TypeIndex::String: {
+            writer->writeString(assert_cast<const ColumnString &>(column).getDataAt(row_num).toView());
+            return;
+        }
+        case TypeIndex::FixedString: {
+            writer->writeString(assert_cast<const ColumnFixedString &>(column).getDataAt(row_num).toView());
+            return;
+        }
+        case TypeIndex::IPv6: {
+            writer->writeIPv6(assert_cast<const ColumnIPv6 &>(column).getElement(row_num));
+            return;
+        }
+        case TypeIndex::Array: {
+            auto nested_type = assert_cast<const DataTypeArray &>(*data_type).getNestedType();
+            const ColumnArray & column_array = assert_cast<const ColumnArray &>(column);
+            const IColumn & nested_column = column_array.getData();
+            const ColumnArray::Offsets & offsets = column_array.getOffsets();
+            size_t offset = offsets[row_num - 1];
+            size_t size = offsets[row_num] - offset;
+            writer->writeStartList();
+            for (size_t i = 0; i < size; ++i)
+                serializeField(nested_column, nested_type, offset + i);
+            writer->writeFinishList();
+            return;
+        }
+        case TypeIndex::Tuple: {
+            const auto & tuple_type = assert_cast<const DataTypeTuple &>(*data_type);
+            const auto & nested_types = tuple_type.getElements();
+            const ColumnTuple & column_tuple = assert_cast<const ColumnTuple &>(column);
+            const auto & nested_columns = column_tuple.getColumns();
+            writer->writeStartList();
+            for (size_t i = 0; i < nested_types.size(); ++i)
+                serializeField(*nested_columns[i], nested_types[i], row_num);
+            writer->writeFinishList();
+            return;
+        }
```

Comment:
Why `Tuple` is serialized as `List` and not as `Struct`?

Reply:
According to the ion [specification](https://amazon-ion.github.io/ion-docs/docs/spec.html#struct), `struct` imply a collection of key/value pairs, I don't understand what key should be then in the case of `Tuple`, so I chose `list`, which implements an ordered list of elements of an arbitrary type.

Reply:
In ClickHouse we have 2 types of `Tuple` actually, named and unnamed `Tuple`. Named `Tuple` looks like this: `Tuple(field_name1 Type1, field_name2 Type2, ...)`. And unnamed `Tuple` looks like this: `Tuple(Type1, Type2, ...)`. And we can output named `Tuple` as `Struct` with field names as keys and unnamed `Tuple` as `List`. You can check if `Tuple` is named of not using `tuple_type.haveExplicitNames()`.

We have the same logic for `JSON` format:
```sql
avogar-dev :) select tuple(1, 'str')::Tuple(field1 UInt64, field2 String) as named_tuple, tuple(1, 'str')::Tuple(UInt64, String) as unnamed_tuple format PrettyJSONEachRow

SELECT
    CAST((1, 'str'), 'Tuple(field1 UInt64, field2 String)') AS named_tuple,
    CAST((1, 'str'), 'Tuple(UInt64, String)') AS unnamed_tuple
FORMAT PrettyJSONEachRow

Query id: 9bbd8cb8-10eb-41d9-aae5-cac291b7be50

{
    "named_tuple": {
        "field1": "1",
        "field2": "str"
    },
    "unnamed_tuple": [
        "1",
        "str"
    ]
}
```